### PR TITLE
Add SHR support for JS-WAM bridge

### PR DESCRIPTION
**Pull request by sameerag**
This feature adds the below support to the token binding work on MSAL Browser:

- Adds support for POP tokens
- Generates SHR in the `msal-browser` in case native layer doesn't support it

### CODE REVIEW

File: lib/msal-browser/src/interaction_client/NativeInteractionClient.ts [hunk 1]
```diff
@@ -236,6 +265,22 @@ export class NativeInteractionClient extends BaseInteractionClient {
 
         const instanceAware: boolean = !!(request.extraQueryParameters && request.extraQueryParameters.instance_aware);
 
+        // validate SHR parameters
+        if (request.authenticationScheme === AuthenticationScheme.POP) {
+            if (!request.resourceRequestMethod && !request.resourceRequestUri) {
```
Comment (jasonnutter) [inline lib/msal-browser/src/interaction_client/NativeInteractionClient.ts]: I thought these were technically optional?

Comment (jasonnutter) [reply lib/msal-browser/src/interaction_client/NativeInteractionClient.ts]: Or does wam shr require them?

**Review by jasonnutter — COMMENTED** [review]
Are there tests that be added? Or will those be added later?

File: lib/msal-browser/src/interaction_client/NativeInteractionClient.ts [hunk 2]
```diff
@@ -24,16 +24,18 @@ import { BrowserAuthError } from "../error/BrowserAuthError";
 export class NativeInteractionClient extends BaseInteractionClient {
     protected apiId: ApiId;
     protected nativeMessageHandler: NativeMessageHandler;
+    protected cryptoObj: ICrypto;
 
     constructor(config: BrowserConfiguration, browserStorage: BrowserCacheManager, browserCrypto: ICrypto, logger: Logger, eventHandler: EventHandler, navigationClient: INavigationClient, apiId: ApiId, provider: NativeMessageHandler, correlationId?: string) {
         super(config, browserStorage, browserCrypto, logger, eventHandler, navigationClient, provider, correlationId);
         this.apiId = apiId;
         this.nativeMessageHandler = provider;
+        this.cryptoObj = browserCrypto;
```
Comment (tnorling) [inline lib/msal-browser/src/interaction_client/NativeInteractionClient.ts]: This is already a part of `BaseInteractionClient` as `this.browserCrypto`

File: lib/msal-common/src/client/AuthorizationCodeClient.ts [hunk 3]
```diff
@@ -404,6 +404,21 @@ export class AuthorizationCodeClient extends BaseClient {
             parameterBuilder.addExtraQueryParameters(request.extraQueryParameters);
         }
 
+        if (request.nativeBroker) {
```
Comment (tnorling) [inline lib/msal-common/src/client/AuthorizationCodeClient.ts:407]: This is also present on line 422 - can we combine?

File: lib/msal-browser/src/interaction_client/NativeInteractionClient.ts (hunk 1 above)
Comment (sameerag) [reply lib/msal-browser/src/interaction_client/NativeInteractionClient.ts]: SHR always required them as far as I can check. Only `shrClaims` and `shrNonce` are optional.

Comment (sameerag) [reply lib/msal-browser/src/interaction_client/NativeInteractionClient.ts]: So yes, in future, when WAM will calculate the SHR, we need to mandate them. We never needed to before, as these never left the browser layer.

**Comment by sameerag** [comment]
> Are there tests that be added? Or will those be added later?

Later. Thomas is manually testing this. I will have a cache PR out soon which will include some unit tests for SHR too.

File: lib/msal-common/src/client/AuthorizationCodeClient.ts (hunk 3 above)
Comment (sameerag) [reply lib/msal-common/src/client/AuthorizationCodeClient.ts:407]: Done, merge issue. Thanks for catching this.

File: lib/msal-browser/src/interaction_client/NativeInteractionClient.ts (hunk 2 above)
Comment (sameerag) [reply lib/msal-browser/src/interaction_client/NativeInteractionClient.ts]: Yes. Edited.

File: lib/msal-browser/src/interaction_client/NativeInteractionClient.ts (hunk 1 above)
Comment (sameerag) [reply lib/msal-browser/src/interaction_client/NativeInteractionClient.ts]: This is from the docs, the code does not restrict anything as such. @hectormmg let's chat about this. I am also following up with WAM team on what they would like to mandate in the future. For now, my preference is to keep this and do an additional PR later with this data to proceed with a single branch testing for WAM JS Bridge.

Comment (sameerag) [reply lib/msal-browser/src/interaction_client/NativeInteractionClient.ts]: Reverting this, the decision is pending based on WAM response, for now, keeping the public surface intact.

**Review by tnorling — APPROVED** [review]

**Pull request merged**